pkg/batchjob/stat.go

```diff
@@ -0,0 +1,226 @@
+package batchjob
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/robfig/cron"
+
+	"github.com/dyweb/sundial/pkg/models"
+	"github.com/dyweb/sundial/pkg/store/rdb"
+	"github.com/dyweb/sundial/pkg/store/tsdb"
+)
+
+//AddCronJob returns a function that could be passed to the cron library.
+func AddCronJob(rdbStore rdb.Store, tsStore tsdb.Store, statRange models.StatRange) func(*cron.Cron) {
+	job := func() {
+		now := time.Now()
+		err := StatWorker(rdbStore, tsStore, now, statRange)
+		if err != nil {
+			fmt.Printf("Error during stat cron at time %s: %s\n", now, err)
+		}
+	}
+	return func(c *cron.Cron) {
+		c.Schedule(
+			cron.Every(ToDuration(statRange)),
+			cron.FuncJob(job))
+	}
+}
+
+//StatWorker collects heartbeats from tsdb and writes stat to rdb
+//collects heartbeats from `startTime` with a given range length
+func StatWorker(rdbStore rdb.Store, tsStore tsdb.Store, beginTime time.Time, statRange models.StatRange) error {
+	hbs, err := tsStore.QueryHeartBeats(beginTime, beginTime.Add(ToDuration(statRange)))
+	if err != nil {
+		return err
+	}
+	stat, err := calculateStat(hbs, beginTime, statRange)
+	if err != nil {
+		return err
+	}
+	err = rdbStore.CreateStat(stat)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//aggregation rules:
+//each heartbeat contains 1 jiffy, whose size is now hardcoded into 1 minute.
+//if two heartbeats are closer than 1 jiffy, merge them.
+//a heartbeat contributes to its project, language, editor, os and dependencies.
+//assume `hbs` is in ascending time order.
+func calculateStat(hbs []models.HeartBeatFrontModel, beginTime time.Time, statRange models.StatRange) (*models.Stat, error) {
+	//TODO: make jiffy configurable, e.g. argument of the cmd
+	jiffy := 60.0 //seconds for one minute
+
+	stat := models.Stat{}
+	// part 1: hardcoded fields
+	stat.IsAlreadyUpdating = false
+	stat.IsCodingActivityVisible = true
+	stat.IsOtherUsageVisible = true
+	stat.IsStuck = false
+	stat.IsUpToDate = true
+	stat.Status = "ok"
+	// these 3 fields are set empty for now, see [process.md](docs/process.md)
+	stat.Editors = []models.Workunit{}
+	stat.OperatingSystems = []models.Workunit{}
+	stat.Dependencies = []models.Workunit{}
+	// part 2: fields that should be read from User
+	//TODO: fix them when we introduce a user
+	stat.Timeout = 60
```

Comment:
so timeout is a per user config instead of global config?

Reply:
yes, timeout means the user's CLI will wait that long from server and decides that there is a connection failure. Configurable by user.